README.rst
*unittest_expander* is a Python library that provides flexible and
easy-to-use tools to parametrize your unit tests, especially (but
not limited to) those based on *unittest.TestCase*.

The library is compatible with Python 3.11, 3.10, 3.9, 3.8, 3.7, 3.6 and
2.7, and does not depend on any external packages (i.e., uses only the
Python standard library).

:Authors: `Jan Kaliszewski (zuo)`_ and `others...`_
:License: `MIT License`_
:Home Page: https://github.com/zuo/unittest_expander
:Documentation: https://unittest-expander.readthedocs.io/en/stable/

.. _Jan Kaliszewski (zuo): https://github.com/zuo/
.. _others...: https://github.com/zuo/unittest_expander/pulls?q=is%3Apr+is%3Amerged
.. _MIT License: https://github.com/zuo/unittest_expander/blob/main/LICENSE.txt


Installing
----------

The easiest way to install the library is to execute (preferably in a
`virtualenv`_) the command::

    python -m pip install unittest_expander

(note that you need network access to do it this way).  If you do not
have the *pip* tool installed -- see:
https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/

Alternatively, you can `download`_ the library source archive, unpack
it, ``cd`` to the unpacked directory and execute (preferably in a
`virtualenv`_) the following command::

    python -m pip install .

Note: you may need to have administrator privileges if you do *not*
operate in a *virtualenv*.

It is also possible to use the library without installing it: as its
code is contained in a single file (``unittest_expander.py``), you can
just copy it into your project.

.. _virtualenv: https://packaging.python.org/tutorials/installing-packages/#creating-and-using-virtual-environments

.. _download: https://pypi.org/project/unittest_expander/#files


Usage example
-------------

Consider the following **ugly** test:

.. code:: python

    import unittest

    class Test(unittest.TestCase):
        def test_sum(self):
            for iterable, expected in [
                ([], 0),
                ([0], 0),
                ([3], 3),
                ([1, 3, 1], 5),
                (frozenset({1, 3}), 4),
                ({1:'a', 3:'b'}, 4),
            ]:
                self.assertEqual(sum(iterable), expected)

Is it cool?  **Not at all!**  So let's improve it:

.. code:: python

    import unittest
    from unittest_expander import expand, foreach

    @expand
    class Test(unittest.TestCase):
        @foreach(
            ([], 0),
            ([0], 0),
            ([3], 3),
            ([1, 3, 1], 5),
            (frozenset({1, 3}), 4),
            ({1:'a', 3:'b'}, 4),
        )
        def test_sum(self, iterable, expected):
            self.assertEqual(sum(iterable), expected)

Now you have **6 distinct tests** (properly *isolated* and being
always *reported as separate tests*), although they share the same
test method source.

You may want to do the same in a bit more verbose and descriptive
way:

.. code:: python

    import unittest
    from unittest_expander import expand, foreach, param

    @expand
    class Test(unittest.TestCase):

        test_sum_params = [
            param([], expected=0).label('empty gives 0'),
            param([0], expected=0),
            param([3], expected=3),
            param([1, 3, 1], expected=5),
            param(frozenset({1, 3}), expected=4),
            param({1:'a', 3:'b'}, expected=4).label('even dict is ok'),
        ]

        @foreach(test_sum_params)
        def test_sum(self, iterable, expected):
            self.assertEqual(sum(iterable), expected)

This is only a fraction of the possibilities *unittest_expander*
offers to you.

You can **learn more** from the actual `documentation of the module
<https://unittest-expander.readthedocs.io/en/stable/narrative_documentation.html>`_.


CHANGES.rst
Changes
=======

0.4.4 (2023-03-21)
------------------

* Documentation: an important completion to the *Narrative
  Documentation* part (regarding the deprecation introduced
  in the previous version of *unittest_expander*) as well as
  several minor improvements and fixes.


0.4.3 (2023-03-21)
------------------

* **Deprecation notice:** using an instance of the Python 3 built-in
  type **bytes** or **bytearray** (or of any subclass thereof) as
  a *parameter collection* -- passed as the *sole* argument to
  **foreach()** or **paramseq()**, or added (using ``+``) to an
  existing **paramseq** object -- is now deprecated (causing emission
  of a **DeprecationWarning**) and will become **illegal** in
  *unittest_expander 0.5.0*.

* A minor fix regarding CI and package metadata...

* Documentation and code comments: minor updates/cleanups.


0.4.2 (2023-03-18)
------------------

* A minor interface usability fix: from now on, the **expand()**
  decorator's attributes **global_name_pattern** and
  **global_name_formatter** are both initially set to **None**
  (previously, by default, they were not initialized at all, so
  trying to get any of them without first setting it caused an
  **AttributeError**).

* Documentation: several updates, improvements and minor fixes.


0.4.1 (2023-03-17)
------------------

* Added to the **unittest_expander** module the **__version__** constant.

* Improvements and additions related to tests, CI, generation of
  documentation, etc.; in particular: added a script that checks whether
  **unittest_expander.__version__** is equal to **version** in package
  metadata, and added invocation of that script to the *Install and
  Test* GitHub workflow.

* Documentation: improvements and minor fixes.


0.4.0 (2023-03-16)
------------------

* From now on, the following versions of Python *are officially
  supported:* **3.11**, **3.10**, **3.9**, **3.8**, **3.7**, **3.6**
  and **2.7** (still).  This means, in particular, that the versions
  *3.5*, *3.4*, *3.3*, *3.2* and *2.6* are *no longer supported*.

* Now, if two (or more) parameter collections are combined to make the
  Cartesian product of them (as an effect of decorating a test with
  two or more **foreach(...)** invocations), and a conflict is detected
  between any *keyword arguments* passed earlier to the **param()**
  constructor to create the **param** instances that are being combined,
  the **expand()** decorator raises a **ValueError** (in older versions
  of *unittest_expander* no exception was raised; instead, a keyword
  argument being a component of one **param** was silently overwritten
  by the corresponding keyword argument being a component of the other
  **param**; that could lead to silent bugs in your tests...).

* When it comes to **param.context()** (and **paramseq.context()**),
  now the standard Python context manager's mechanism of suppressing
  exceptions (by making **__exit__()** return a *true* value) is,
  by default, consistently *disabled* (i.e. exceptions are *not*
  suppressed, as it could easily become a cause of silent test bugs; the
  previous behavior was inconsistent: exceptions could be suppressed in
  the case of applying **foreach()** decorators to test *methods*, but
  not in the case of applying them to test *classes*).

  If needed, the possibility of suppressing exceptions by **__exit__()**
  returning a *true* value can be explicitly *enabled* on a case-by-case
  basis by passing ``_enable_exc_suppress_=True`` to **param.context()**
  (or **paramseq.context()**).

* **Deprecation notice:** decorating test *classes* with **foreach()**
  -- to generate new parametrized test *classes* -- is now deprecated
  (causing emission of a **DeprecationWarning**); in future versions of
  *unittest_expander* it will first become **unsupported** (in *0.5.0*),
  and then, in some version, will (most probably) get a **new meaning**.
  The current shape of the feature is deemed broken by design (in
  particular, because of the lack of composability; that becomes
  apparent when class inheritance comes into play...).

* **Deprecation notice:** a change related to the deprecation described
  above is that now the **expand()**'s keyword argument ``into`` is also
  deprecated (its use causes emission of a **DeprecationWarning**) and
  will become **illegal** in *unittest_expander 0.5.0*.

* **Deprecation notice:** using a tuple (i.e., an instance of the
  built-in type **tuple** or of any subclass of it, e.g., a *named
  tuple*) as a *parameter collection* -- passed as the *sole* argument
  to **foreach()** or **paramseq()**, or added (using ``+``) to an
  existing **paramseq** object -- is now deprecated (causing emission
  of a **DeprecationWarning**) and will become **illegal** in
  *unittest_expander 0.5.0*.  Instead of a tuple, use a collection
  of another type (e.g., a **list**).

  Note: this deprecation concerns tuples used as *parameter collections*,
  *not* as *items* of parameter collections (tuples being such items,
  acting as simple substitutes of **param** objects, are -- and will
  always be -- perfectly OK).

* Two compatibility fixes:

  (1) now test methods with *keyword-only* arguments and/or *type
  annotations* are supported (previously an error was raised by
  **expand()** if such a method was decorated with **foreach()**);
  the background is that under Python 3.x, from now on,
  **inspect.getfullargspec()** (instead of its legacy predecessor,
  **inspect.getargspec()**) is used to inspect test method signatures;

  (2) now standard *abstract base classes* of collections are imported

unittest_expander/_internal.py
# Copyright (c) 2014 Jan Kaliszewski (zuo). All rights reserved.
# Licensed under MIT License (see the LICENSE.txt file for details).

import collections
import functools
import inspect
import itertools
import string
import sys
import types


_PY3 = sys.version_info[0] >= 3
_CLASS_TYPES = (type,) if _PY3 else (type, types.ClassType)
_STRING_TYPES = (str,) if _PY3 else (str, unicode)

_PARAMSEQ_OBJS_ATTR = '__attached_paramseq_objs'

_GENERIC_KWARGS = 'context_targets', 'label'

_DEFAULT_PARAMETRIZED_NAME_PATTERN = '{base_name}__<{label}>'
_DEFAULT_PARAMETRIZED_NAME_FORMATTER = string.Formatter()


class _Context(object):

    def __init__(self, context_manager_factory, *args, **kwargs):
        self._context_manager_factory = context_manager_factory
        self._args = args
        self._kwargs = kwargs

    def _make_context_manager(self):
        return self._context_manager_factory(*self._args, **self._kwargs)


class Substitute(object):

    def __init__(self, actual_object):
        self.actual_object = actual_object

    def __getattribute__(self, name):
        if name in ('actual_object', '__class__', '__call__'):
            return super(Substitute, self).__getattribute__(name)
        return getattr(self.actual_object, name)

    def __dir__(self):
        names = ['actual_object']
        names.extend(
            name
            for name in dir(self.actual_object)
            if name not in ('actual_object', '__call__'))
        return names


class param(object):

    def __init__(self, *args, **kwargs):
        self._args = args
        self._kwargs = kwargs
        self._context_list = []
        self._label_list = []

    def context(self, context_manager_factory, *args, **kwargs):
        context = _Context(context_manager_factory, *args, **kwargs)
        return self._from_components(
            self._args,
            self._kwargs,
            self._context_list + [context],
            self._label_list)

    def label(self, text):
        return self._from_components(
            self._args,
            self._kwargs,
            self._context_list,
            self._label_list + [text])

    @classmethod
    def _from_param_item(cls, param_item):
        if isinstance(param_item, param):
            return param_item
        if isinstance(param_item, tuple):
            return cls(*param_item)
        return cls(param_item)

    @classmethod
    def _combine_instances(cls, param_instances):
        args = []
        kwargs = {}
        context_list = []
        label_list = []
        for param_inst in param_instances:
            args.extend(param_inst._args)
            kwargs.update(param_inst._kwargs)
            context_list.extend(param_inst._context_list)
            label_list.append(param_inst._get_label())
        return cls._from_components(args, kwargs, context_list, label_list)

    @classmethod
    def _from_components(cls, args, kwargs, context_list, label_list):
        new = cls(*args, **kwargs)
        new._context_list.extend(context_list)
        new._label_list.extend(label_list)
        return new

    def _get_context_manager_factory(self):
        try:
            return self.__cached_cm_factory
        except AttributeError:
            if self._context_list:
                # we need to combine several context managers (from the
                # contexts) but Python 2 does not have contextlib.ExitStack
                # and contextlib.nested() is deprecated (for good reasons)
                # -- so we will generate, compile and exec the code:
                src_code = (
                    'import contextlib\n'
                    '@contextlib.contextmanager\n'
                    'def cm_factory():\n'
                    '    context_targets = [None] * len(context_list)\n'
                    '    {0}'
                    'yield context_targets\n'.format(''.join(
                        ('with context_list[{0}]._make_context_manager() '
                         'as context_targets[{0}]:\n{next_indent}'
                        ).format(i, next_indent=((8 + 4 * i) * ' '))
                        for i in range(len(self._context_list)))))
                # Py2+Py3-compatibile substitute of exec in a given namespace
                code = compile(src_code, '<string>', 'exec')
                namespace = {'context_list': self._context_list}
                eval(code, namespace)
                self.__cached_cm_factory = namespace['cm_factory']
            else:
                self.__cached_cm_factory = None
            return self.__cached_cm_factory

    def _get_label(self):
        if self._label_list:
            return ', '.join(label for label in self._label_list)
        else:
            short_repr = self._short_repr
            args_reprs = (short_repr(val) for val in self._args)
            kwargs_reprs = ('{0}={1}'.format(key, short_repr(val))
                            for key, val in sorted(self._kwargs.items()))
            return ','.join(itertools.chain(args_reprs, kwargs_reprs))

    @staticmethod
    def _short_repr(obj, max_len=16):
        r = repr(obj)
        if len(r) > max_len:
            r = '<{0}...>'.format(r.lstrip('<')[:max_len-5])
        return r


class paramseq(object):

    def __init__(*posargs, **some_param_source_dict):
        try:
            self, some_param_source = posargs
        except ValueError:
            try:
                [self] = posargs
            except ValueError:
                raise TypeError(
                    '__init__() takes 1 or 2 positional arguments '
                    '({0} given)'.format(len(posargs)))
            self._init_with_param_sources(some_param_source_dict)
        else:
            if some_param_source_dict:
                raise TypeError(
                    'paramseq constructor can take either one positional '
                    'argument or some keyword arguments, not both')
            else:
                self._init_with_param_sources(some_param_source)

    def __add__(self, other):
        if self._is_instance_of_legal_param_source_class(other):
            return self._from_param_sources(self, other)
        return NotImplemented

    def __radd__(self, other):
        if self._is_instance_of_legal_param_source_class(other):
            return self._from_param_sources(other, self)
        return NotImplemented

    def context(self, context_manager_factory, *args, **kwargs):
        context = _Context(context_manager_factory, *args, **kwargs)
        new = self._from_param_sources(self)
        new._context_list.append(context)
        return new

    @classmethod
    def _from_param_sources(cls, *param_sources):
        self = cls.__new__(cls)
        self._init_with_param_sources(*param_sources)
        return self

    def _init_with_param_sources(self, *param_sources):
        param_sources = tuple(param_sources)
        for param_src in param_sources:
            if not self._is_instance_of_legal_param_source_class(param_src):
                raise TypeError(
                    'class {0.__class__!r} (of {0!r}) is not a '
                    'legal parameter source class'.format(param_src))
        self._param_sources = param_sources
        self._context_list = []

    @staticmethod
    def _is_instance_of_legal_param_source_class(obj):
        return isinstance(obj, (
            paramseq,
            collections.Sequence,
            collections.Set,
            collections.Mapping,
            collections.Callable)
        ) and not isinstance(obj, _STRING_TYPES)

    def _generate_params(self, test_cls):
        for param_inst in self._generate_raw_params(test_cls):
            if self._context_list:
                param_inst = param_inst._from_components(
                    param_inst._args,
                    param_inst._kwargs,
                    param_inst._context_list + self._context_list,
                    param_inst._label_list)
            yield param_inst

    def _generate_raw_params(self, test_cls):
        for param_src in self._param_sources:
            if isinstance(param_src, paramseq):
                for param_inst in param_src._generate_params(test_cls):
                    yield param_inst
            elif isinstance(param_src, collections.Mapping):
                for label, param_item in param_src.items():
                    yield param._from_param_item(param_item).label(label)
            else:
                if isinstance(param_src, collections.Callable):
                    try:
                        param_src = param_src(test_cls)
                    except TypeError:
                        param_src = param_src()
                else:
                    assert isinstance(param_src, (collections.Sequence,
                                                  collections.Set))
                for param_item in param_src:
                    yield param._from_param_item(param_item)


# test case *method* or *class* decorator...
def foreach(*posargs, **some_param_source_dict):
    ps = paramseq(*posargs, **some_param_source_dict)
    def decorator(func_or_cls):
        stored_paramseq_objs = getattr(func_or_cls, _PARAMSEQ_OBJS_ATTR, None)
        if stored_paramseq_objs is None:
            stored_paramseq_objs = []
            setattr(func_or_cls, _PARAMSEQ_OBJS_ATTR, stored_paramseq_objs)
        assert isinstance(stored_paramseq_objs, list)
        stored_paramseq_objs.append(ps)
        return func_or_cls
    return decorator


# test case *class* decorator...
def expand(test_cls=None, **kwargs):
    into = kwargs.pop('into', None)
    if kwargs:
        raise TypeError(
            'expand() got unexpected keyword arguments: ' +
            ', '.join(sorted(map(repr, kwargs))))
    if test_cls is None:
        return functools.partial(expand, into=into)
    _expand_test_methods(test_cls)
    return _expand_test_cls(test_cls, into)


def _expand_test_methods(test_cls):
    attr_names = dir(test_cls)
    seen_names = set(attr_names)
    attrs_to_substitute = dict()
    attrs_to_add = dict()
    for base_name in attr_names:
        obj = getattr(test_cls, base_name, None)
        paramseq_objs = getattr(obj, _PARAMSEQ_OBJS_ATTR, None)
        if paramseq_objs is not None:
            if _PY3:
                # no unbound methods in Python 3
                if not isinstance(obj, types.FunctionType):
                    raise TypeError('{0!r} is not a function'.format(obj))
                base_func = obj
            else:
                if not isinstance(obj, types.MethodType):
                    raise TypeError('{0!r} is not a method'.format(obj))
                base_func = obj.__func__
            arg_spec = inspect.getargspec(base_func)
            accepted_generic_kwargs = set(
                _GENERIC_KWARGS if arg_spec.keywords is not None
                else (kw for kw in _GENERIC_KWARGS
                      if kw in arg_spec.args))
            for func in _generate_parametrized_functions(
                    test_cls, paramseq_objs,
                    base_name, base_func, seen_names,
                    accepted_generic_kwargs):
                attrs_to_add[func.__name__] = func
            attrs_to_substitute[base_name] = obj
    for name, obj in attrs_to_substitute.items():
        setattr(test_cls, name, Substitute(obj))
    for name, obj in attrs_to_add.items():
        setattr(test_cls, name, obj)


def _expand_test_cls(base_test_cls, into):
    paramseq_objs = getattr(base_test_cls, _PARAMSEQ_OBJS_ATTR, None)
    if paramseq_objs is None:
        return base_test_cls
    else:
        if not isinstance(base_test_cls, _CLASS_TYPES):
            raise TypeError('{0!r} is not a class'.format(base_test_cls))
        into = _resolve_the_into_arg(into, globals_frame_depth=3)
        seen_names = set(list(into.keys()) + [base_test_cls.__name__])
        for cls in _generate_parametrized_classes(
                base_test_cls, paramseq_objs, seen_names):
            into[cls.__name__] = cls
        return Substitute(base_test_cls)


def _resolve_the_into_arg(into, globals_frame_depth):
    orig_into = into
    if into is None:
        into = sys._getframe(globals_frame_depth).f_globals['__name__']
    if isinstance(into, _STRING_TYPES):
        into = __import__(into, globals(), locals(), ['*'], 0)
    if inspect.ismodule(into):
        into = vars(into)
    if not isinstance(into, collections.MutableMapping):
        raise TypeError(
            "resolved 'into' argument is not a mutable mapping "
            "({0!r} given, resolved to {1!r})".format(orig_into, into))
    return into


def _generate_parametrized_functions(test_cls, paramseq_objs,
                                     base_name, base_func, seen_names,
                                     accepted_generic_kwargs):
    for count, param_inst in enumerate(
            _generate_params_from_sources(paramseq_objs, test_cls),
            start=1):
        yield _make_parametrized_func(base_name, base_func, count, param_inst,
                                      seen_names, accepted_generic_kwargs)


def _generate_parametrized_classes(base_test_cls, paramseq_objs, seen_names):
    for count, param_inst in enumerate(
            _generate_params_from_sources(paramseq_objs, base_test_cls),
            start=1):
        yield _make_parametrized_cls(base_test_cls, count,
                                     param_inst, seen_names)


def _generate_params_from_sources(paramseq_objs, test_cls):
    src_params_iterables = [
        ps._generate_params(test_cls)
        for ps in paramseq_objs]
    for params_row in itertools.product(*src_params_iterables):
        yield param._combine_instances(params_row)


def _make_parametrized_func(base_name, base_func, count, param_inst,
                            seen_names, accepted_generic_kwargs):
    p_args = param_inst._args
    p_kwargs = param_inst._kwargs
    label = param_inst._get_label()
    cm_factory = param_inst._get_context_manager_factory()
    if cm_factory is None:
        @functools.wraps(base_func)
        def generated_func(*args, **kwargs):
            args += p_args
            kwargs.update(**p_kwargs)
            if 'context_targets' in accepted_generic_kwargs:
                kwargs.setdefault('context_targets', [])
            if 'label' in accepted_generic_kwargs:
                kwargs.setdefault('label', label)
            return base_func(*args, **kwargs)
    else:
        @functools.wraps(base_func)
        def generated_func(*args, **kwargs):
            args += p_args
            kwargs.update(**p_kwargs)
            with cm_factory() as context_targets:
                if 'context_targets' in accepted_generic_kwargs:
                    kwargs.setdefault('context_targets', context_targets)
                if 'label' in accepted_generic_kwargs:
                    kwargs.setdefault('label', label)
                return base_func(*args, **kwargs)
    generated_func.__name__ = _format_name_for_parametrized(
        base_name, base_func, label, count, seen_names)
    _set_qualname(base_func, generated_func)
    return generated_func


def _make_parametrized_cls(base_test_cls, count, param_inst, seen_names):
    cm_factory = param_inst._get_context_manager_factory()
    label = param_inst._get_label()

    class generated_test_cls(base_test_cls):

        def setUp(self):
            self.label = label
            self.params = param_inst._args
            for name, obj in param_inst._kwargs.items():
                setattr(self, name, obj)
            exit = None
            try:
                if cm_factory is not None:
                    cm = cm_factory()
                    cm_type = type(cm)
                    if not _PY3 and isinstance(cm_type, types.InstanceType):
                        # old-style class (Python 2 only)
                        cm_exit = cm.__exit__
                        cm_enter = cm.__enter__
                        self.context_targets = cm_enter()
                        exit = cm_exit
                    else:
                        # new-style class
                        cm_type_exit = cm_type.__exit__
                        cm_type_enter = cm_type.__enter__
                        self.context_targets = cm_type_enter(cm)
                        def exit(*exc_info):
                            return cm_type_exit(cm, *exc_info)
                self.__exit = exit
                return super(generated_test_cls, self).setUp()
            except:
                if exit is not None:
                    exc_info = sys.exc_info()
                    try:
                        exit(*exc_info)
                    finally:
                        self.__exit = None
                raise

        def tearDown(self):
            try:
                r = super(generated_test_cls, self).tearDown()
            except:
                exc_info = sys.exc_info()
                exit = self.__exit
                if exit is not None:
                    exit(*exc_info)
                raise
            else:
                exit = self.__exit
                if exit is not None:
                    exit(None, None, None)
                return r
            finally:
                self.__exit = None

    generated_test_cls.__module__ = base_test_cls.__module__
    generated_test_cls.__name__ = _format_name_for_parametrized(
        base_test_cls.__name__, base_test_cls, label, count, seen_names)
    _set_qualname(base_test_cls, generated_test_cls)
    return generated_test_cls


def _format_name_for_parametrized(base_name, base_obj,
                                  label, count, seen_names):
    pattern, formatter = _get_name_pattern_and_formatter()
    name = stem_name = formatter.format(
        pattern,
        base_name=base_name,
        base_obj=base_obj,
        label=label,
        count=count)
    uniq_tag = 2
    while name in seen_names:
        # ensure that, for a particular class, names are unique
        name = '{0}__{1}'.format(stem_name, uniq_tag)
        uniq_tag += 1
    seen_names.add(name)
    return name


def _get_name_pattern_and_formatter():
    pattern = getattr(expand, 'global_name_pattern', None)
    if pattern is None:
        pattern = _DEFAULT_PARAMETRIZED_NAME_PATTERN
    formatter = getattr(expand, 'global_name_formatter', None)
    if formatter is None:
        formatter = _DEFAULT_PARAMETRIZED_NAME_FORMATTER
    return pattern, formatter


def _set_qualname(base_obj, target_obj):
    # relevant to Python 3.3+
    base_qualname = getattr(base_obj, '__qualname__', None)
    if base_qualname is not None:
        base_name = base_obj.__name__
        qualname_prefix = (
            base_qualname[:(len(base_qualname) - len(base_name))]
            if (base_qualname == base_name or
                base_qualname.endswith('.' + base_name))
            else '<...>.')
        target_obj.__qualname__ = qualname_prefix + target_obj.__name__
